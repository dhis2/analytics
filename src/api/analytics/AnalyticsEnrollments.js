import AnalyticsBase from './AnalyticsBase.js'

/**
 * @extends module:analytics.AnalyticsBase
 *
 * @description
 * Analytics enrollments class used to request analytics enrollments data from Web API.
 *
 * @memberof module:analytics
 *
 * @see https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-236/analytics.html#webapi_enrollment_analytics
 */
class AnalyticsEnrollments extends AnalyticsBase {
    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics query data from the api.
     *
     * @example
     * const req = new analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  analytics.enrollments.getQuery(req)
     *  .then(console.log);
     */
    getQuery(req) {
        return this.fetch(req.withPath('enrollments/query'))
    }
}

export default AnalyticsEnrollments
