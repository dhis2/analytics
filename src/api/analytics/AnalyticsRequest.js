import { getExpressionHashFromVisualization } from '../../modules/hash.js'
import { getFixedDimensions } from '../../modules/predefinedDimensions.js'
import AnalyticsRequestBase from './AnalyticsRequestBase.js'
import AnalyticsRequestDimensionsMixin from './AnalyticsRequestDimensionsMixin.js'
import AnalyticsRequestFiltersMixin from './AnalyticsRequestFiltersMixin.js'
import AnalyticsRequestPropertiesMixin from './AnalyticsRequestPropertiesMixin.js'
import { formatDimension } from './utils.js'

/**
 * @description
 * Class for constructing a request object to use for communicating with the analytics API endpoint.
 *
 * @param {!Object} options Object with analytics request options
 *
 * @memberof module:analytics
 *
 * @extends module:analytics.AnalyticsRequestDimensionsMixin
 * @extends module:analytics.AnalyticsRequestFiltersMixin
 * @extends module:analytics.AnalyticsRequestPropertiesMixin
 * @extends module:analytics.AnalyticsRequestBase
 */
class AnalyticsRequest extends AnalyticsRequestDimensionsMixin(
    AnalyticsRequestFiltersMixin(
        AnalyticsRequestPropertiesMixin(AnalyticsRequestBase)
    )
) {
    /**
     * Extracts dimensions and filters from an analytic object visualization and add them to the request
     *
     * @param {Object} visualization The analytics object visualization from which extract the dimensions/filters
     * @param {Boolean} [passFilterAsDimension=false] Pass filters as dimension in the query string (used in dataValueSet requests)
     *
     * @returns {AnalyticsRequest} A new instance of the class for chaining purposes
     *
     * @example
     * const req = new analytics.request()
     *    .fromVisualization(visualization);
     *
     * // dimension=pe:last_12_month&dimension=dx:fbfjhsppuqd;cyeuwxtcpku;jtf34knzhzp;hfdmmspbglg&filter=ou:imsptqpwcqd
     *
     * const req2 = new analytics.request()
     *    .fromvisualization(visualization, true);
     *
     * // dimension=pe:last_12_month&dimension=dx:fbfjhsppuqd;cyeuwxtcpku;jtf34knzhzp;hfdmmspbglg&dimension=ou:imsptqpwcqd
     */
    fromVisualization(visualization, passFilterAsDimension = false) {
        let request = this

        const outputType = visualization.outputType

        // extract dimensions from visualization
        const columns = visualization.columns || []
        const rows = visualization.rows || []

        columns.concat(rows).forEach((d) => {
            let dimension = d.dimension

            if (d.legendSet?.id) {
                dimension += `-${d.legendSet.id}`
            }

            if (d.filter) {
                dimension += `:${d.filter}`
            }

            const programStageId = d.programStage?.id

            if (d.repetition?.indexes?.length) {
                d.repetition.indexes.forEach((index) => {
                    request = request.addDimension(
                        formatDimension({
                            programId: d.program?.id,
                            programStageId: `${programStageId}[${index}]`,
                            dimension,
                            outputType,
                        })
                    )
                })
            } else {
                request = request.addDimension(
                    formatDimension({
                        programId: d.program?.id,
                        programStageId,
                        dimension,
                        outputType,
                    }),
                    d.items?.map((item) => item.id)
                )
            }
        })

        // extract filters from visualization
        const filters = visualization.filters || []

        // only pass dx/pe/ou as dimension
        const fixedIds = Object.keys(getFixedDimensions())

        filters.forEach((f) => {
            if (passFilterAsDimension && fixedIds.includes(f.dimension)) {
                request = request.addDimension(
                    f.dimension,
                    f.items?.map((item) => item.id)
                )
            } else {
                let filterString = f.dimension

                if (f.filter) {
                    filterString += `:${f.filter}`
                }

                const programStageId = f.programStage?.id

                if (f.repetition?.indexes?.length) {
                    f.repetition.indexes.forEach((index) => {
                        request = request.addFilter(
                            formatDimension({
                                programId: f.program?.id,
                                programStageId: `${programStageId}[${index}]`,
                                dimension: filterString,
                                outputType,
                            })
                        )
                    })
                } else {
                    request = request.addFilter(
                        formatDimension({
                            programId: f.program?.id,
                            programStageId,
                            dimension: filterString,
                            outputType,
                        }),
                        f.items?.map((item) => item.id)
                    )
                }
            }
        })

        // add cache param for expression dimension items
        const expressionHash = getExpressionHashFromVisualization(visualization)

        if (expressionHash) {
            request.withParameters({ edi_cache: expressionHash })
        }

        return request
    }
}

export default AnalyticsRequest
