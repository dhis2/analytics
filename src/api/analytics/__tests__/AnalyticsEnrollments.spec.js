import fixtures from '../../../__fixtures__/fixtures.js'
import DataEngineMock from '../__mocks__/DataEngine.js'
import AnalyticsEnrollments from '../AnalyticsEnrollments.js'
import AnalyticsRequest from '../AnalyticsRequest.js'

describe('analytics.enrollments', () => {
    let enrollments
    let request
    let dataEngineMock
    let fixture

    beforeEach(() => {
        dataEngineMock = new DataEngineMock()
        DataEngineMock.mockClear()
        enrollments = new AnalyticsEnrollments()
    })

    it('should not be allowed to be called without new', () => {
        expect(() => AnalyticsEnrollments()).toThrowErrorMatchingSnapshot()
    })

    it('should use the dataEngine object when it is passed', () => {
        const dataEngineMockObject = {}

        enrollments = new AnalyticsEnrollments(dataEngineMockObject)

        expect(enrollments.dataEngine).toBe(dataEngineMockObject)
    })

    describe('.getQuery()', () => {
        beforeEach(() => {
            enrollments = new AnalyticsEnrollments(new DataEngineMock())

            request = new AnalyticsRequest()
                .addOrgUnitDimension('ImspTQPwCqd')
                .addDimension('WZbXY0S00lP.de0FEHSIoxh')
                .addDimension('WZbXY0S00lP.sWoqcoByYmD')
                .addPeriodFilter('LAST_MONTH')
                .withProgram('WSGAb5XwJ3Y')
                .withStage('WZbXY0S00lP')
                .withAsc('ENROLLMENTDATE')
                .withOuMode('DESCENDANTS')
                .withColumns('w75KJ2mc4zz')
                .withPage(1)
                .withPageSize(10)

            fixture = fixtures.get('/api/analytics/enrollments')

            dataEngineMock.query.mockReturnValue(
                Promise.resolve({ data: fixture })
            )
        })

        it('should be a function', () => {
            expect(enrollments.getQuery).toBeInstanceOf(Function)
        })

        it('should resolve a promise with data', () =>
            enrollments.getQuery(request).then((data) => {
                expect(data.width).toEqual(fixture.width)
                expect(data.height).toEqual(fixture.height)
            }))
    })
})
