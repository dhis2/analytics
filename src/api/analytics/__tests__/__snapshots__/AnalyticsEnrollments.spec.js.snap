// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analytics.enrollments should not be allowed to be called without new 1`] = `"Class constructor AnalyticsEnrollments cannot be invoked without 'new'"`;
